File: kotlin/src/main/kotlin/y2022/day03.kt
package y2022

import readLines

fun main() {
    val data = readLines("y2022/day03")
    val letters = ('a'..'z') + ('A'..'Z')

    val compartments = data
        .map { line -> line.chunked(line.length / 2) }
        .map { compartments -> compartments.map { it.toSet() } }
        .map { (c1, c2) -> (c1 intersect c2).first() }
        .sumOf { letters.indexOf(it) + 1 }

    val groups = data
        .chunked(3)
        .map { group -> group.map { it.toSet() } }
        .map { (g1, g2, g3) -> (g1 intersect g2 intersect g3).first() }
        .sumOf { letters.indexOf(it) + 1 }

    println(compartments)
    println(groups)
}

File: 2022/day03.py
from string import ascii_letters

with open("input_files/day03", "r") as f:
    data = f.read().splitlines()

middles = [len(line) // 2 for line in data]
compartments = [(line[:i], line[i:]) for line, i in zip(data, middles)]
compartments = [(set(c1) & set(c2)).pop() for c1, c2 in compartments]

groups = [data[i: i + 3] for i in range(0, len(data), 3)]
groups = [map(set, group) for group in groups]
groups = [set.intersection(*group).pop() for group in groups]

print(sum(ascii_letters.index(item) + 1 for item in compartments))
print(sum(ascii_letters.index(item) + 1 for item in groups))
